constraints/miden-vm/memory.air

```diff
@@ -0,0 +1,104 @@
+mod MemoryChipletAir
+
+### Helper functions/evaluators ###################################################################
+
+fn delta(v: scalar) -> scalar:
+    return v' - v
+
+
+# Enforces that the provided columns must be zero.
+ev is_zero(main: [values[4]]):
+    enf v = 0 for v in values
+
+
+# Enforces that values in the columns in the current row must be equal to the values in the next 
+# row.
+ev is_unchanged(main: [values[4]]):
+    enf v' = v for v in values
+
+
+# Enforces that the provided columns must be binary.
+ev is_binary(main: [a]):
+    enf a^2 - a = 0
+
+
+# Enforces that created flags and s[0] column have valid values during the program execution.
+ev flags_validity(main: [s[2], ctx, addr, t]):
+    # n0 = 1 when context changes and 0 otherwise.
+    let n0 = delta(ctx) * t'
+
+    # if context remains the same, n1 = 1 when address changes and 0 otherwise.
+    let n1 = delta(addr) * t'
+
+    # Enforce that n0 must be binary.
+    enf n0^2 = n0
```

Comment:
Similar to my earlier comment about consistency - in the hash chiplet PR we use `b^2 - b = 0`. I prefer this way of doing it, and would prefer to do it in the same way everywhere. This would suggest changing to `b^2 = b` in the hash chiplet PR